control-plane/agents/src/bin/core/tests/rebuild/mod.rs

```diff
@@ -0,0 +1,217 @@
+        if std::time::Instant::now() > (start + timeout)
+            && (pool.state.is_none() || pool.state.as_ref().unwrap().status != PoolStatus::Online)
+        {
+            return None;
+        }
+        tokio::time::sleep(Duration::from_secs(1)).await;
+    }
+}
+
+// This test
+// 1. creates a volume with 3 replicas, faults one of them by restarting container.
+// 2. Expects child's replica to be online again within 15sec period.
+// 3. Attempts to online the child.
+// 4. Child should get successfully onlined and partial rebuild happen.
+// 5. If online call fails, full rebuild of child happens.
+#[tokio::test]
+async fn replica_wait_and_partial_rebuild() {
+    let cluster = build_cluster(4, 52428800, 0).await; // 50MiB pool size
+
+    let vol_target = cluster.node(0).to_string();
+    let api_client = cluster.rest_v00();
+    let volume_api = api_client.volumes_api();
+
+    let body = CreateVolumeBody::new(VolumePolicy::new(true), 3, 10485760u64, false);
+    let volid = VolumeId::new();
+    let volume = volume_api.put_volume(&volid, body).await.unwrap();
+    let volume = volume_api
+        .put_volume_target(
+            &volume.spec.uuid,
+            PublishVolumeBody::new_all(
+                HashMap::new(),
+                None,
+                vol_target.clone().to_string(),
+                models::VolumeShareProtocol::Nvmf,
+                None,
+                cluster.csi_node(0),
+            ),
+        )
+        .await
+        .unwrap();
+
+    let volume_state = volume.state;
+    assert!(volume_state.status == VolumeStatus::Online.into());
+    let nexus = volume_state.target.unwrap();
+    assert_eq!(nexus.children.len(), 3);
+    nexus
+        .children
+        .iter()
+        .for_each(|c| assert!(c.state == ChildState::Online.into()));
+
+    // Check where the replicas are, apart from vol target node.
+    let replicas = api_client.replicas_api().get_replicas().await.unwrap();
+    tracing::info!({?replicas}, "Available replicas - ");
+    let testrep = replicas.iter().find(|r| r.node != vol_target).unwrap();
+    let testrep_node = &testrep.node;
+    tracing::info!(
+        "Restarting node {testrep_node} having replica {}",
+        testrep.uri.clone()
+    );
+
+    // Restart the container.
+    cluster
+        .composer()
+        .restart(&testrep_node.to_string())
+        .await
+        .expect("container stop failure");
+
+    // Check the pool on restarted node is imported back.
+    wait_till_pool_online(&cluster, testrep, Duration::from_secs(20)).await;
+
+    // Fetch replicas again.
+    let replicas_now = api_client.replicas_api().get_replicas().await.unwrap();
+    // The replica picked for test must be online now again.
+    assert!(replicas_now.iter().any(|r| r.uri == testrep.uri));
+    replicas_now.iter().find(|r| {
+        if r.uri == testrep.uri {
+            assert!(r.state == ReplicaStatus::Online.into());
+            return true;
+        }
+        false
+    });
```

Comment:
You're not using anything from the find, did you mean to use for_each instead?

```suggestion
    replicas_now.iter().for_each(|r| {
        if r.uri == testrep.uri {
            assert!(r.state == ReplicaStatus::Online.into());
        }
    });    
```
or 
```suggestion
    replicas_now.iter().filter(|r| r.uri == testrep.uri).for_each(|r| {
        assert!(r.state == ReplicaStatus::Online.into());
    });
```

Reply:
hmm yeah. preferred `find` because it short-circuits, but yes my usage is clumsy currently. How about this:

```
    if let Some(rep) = replicas_now.iter().find(|r| r.uri == testrep.uri) {
        assert!(rep.state == ReplicaState::Online);
    }
```

Reply:
ok... lgtm